Fix osu! stacking not matching stable in very specific edge cases

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/36052.

Not much more to say here. Until now the `PreEmpt` typing discrepancy has likely gone unnoticed since it's usually only used in visual usages.

peppy committed:
Fix osu! stacking not matching stable in very specific edge case

Closes https://github.com/ppy/osu/issues/36052.

Not much more to say here. Until now the `PreEmpt` typing discrepancy
has likely gone unnoticed since it's usually only used in visual usages.

bdach reviewed (commented):
if we're already tracking this down I would probably apply the truncation in a few other places:

https://github.com/ppy/osu/blob/c1ea472dcf86f2b0f78e2c73aeb77164f8f138da/osu.Game.Rulesets.Catch/Objects/CatchHitObject.cs#L153

this one shouldn't have any material effects but maybe? who knows at this point

https://github.com/ppy/osu/blob/af290ddc0ca0dc51b19154173b8c1ed0358f1028/osu.Game.Rulesets.Osu/OsuRuleset.cs#L415
https://github.com/ppy/osu/blob/af290ddc0ca0dc51b19154173b8c1ed0358f1028/osu.Game.Rulesets.Catch/CatchRuleset.cs#L303

these ones are purely visual, probably best to not pretend fractions are involved here when they're not

peppy committed:
Add helper method and update some other similar preempt calculations

bdach reviewed (dismissed):
pending diffcalc check

(I posted a wrong comment here earlier then deleted it because I confused myself)

bdach committed:
Add extended conversion mapping coverage for more failing cases

Some are not immediately relevant to the stacking issue because they
fail both before and after it, just less so after the stacking issue
(half-)fix, and as such have been commented out for the time being.

bdach committed:
Fix stack threshold being calculated in a different way than stable

bdach commented:
I've pushed a fix for the regressed beatmap. @peppy please check if you're ok with my changes whenever able.

In the meantime:

!diffcalc
RULESET=osu

The other broken cases I'll look at separately.

bdach committed:
Fix yet another incorrect threshold check in stacking code

bdach committed:
Remove now-redundant tests

They all pass along with the last stacking threshold fix, so I see no
reason to keep them in.
